text/0000-ppv.md

```diff
@@ -0,0 +1,1596 @@
+/// reading or dropping the old value.
+///
+/// # Panics
+///
+/// If `slice.len() != Self::lanes()` or `&slice[0]` is not
+/// aligned to an `align_of::<Self>()` boundary.
+pub fn write_aligned(self, slice: &mut [element_type]);
+
+/// Writes the values of the vector to the `slice` without 
+/// reading or dropping the old value.
+///
+/// # Panics
+///
+/// If `slice.len() != Self::lanes()`.
+pub fn write_unaligned(self, slice: &mut [element_type]);
+
+/// Writes the values of the vector to the `slice` without 
+/// reading or dropping the old value.
+///
+/// # Precondition
+///
+/// If `slice.len() < Self::lanes()` or `&slice[0]` is not
+/// aligned to an `align_of::<Self>()` boundary, the behavior is
+/// undefined.
+pub unsafe fn write_aligned_unchecked(self, slice: &mut [element_type]);
+
+/// Writes the values of the vector to the `slice` without reading 
+/// or dropping the old value.
+///
+/// # Precondition
+///
+/// If `slice.len() < Self::lanes()` the behavior is undefined.
+pub unsafe fn write_unaligned_unchecked(self, slice: &mut [element_type]);
+
+/// Instantiates a new vector with the values of the `slice` without 
+/// moving them, leaving the memory in `slice` unchanged.
+///
+/// # Panics
+///
+/// If `slice.len() != Self::lanes()` or `&slice[0]` is not aligned
+/// to an `align_of::<Self>()` boundary.
+pub fn read_aligned(slice: &[element_type]) -> Self;
+
+/// Instantiates a new vector with the values of the `slice` without 
+/// moving them, leaving the memory in `slice` unchanged.
+///
+/// # Panics
+///
+/// If `slice.len() != Self::lanes()`.
+pub fn read_unaligned(slice: &[element_type]) -> Self;
+
+/// Instantiates a new vector with the values of the `slice` without 
+/// moving them, leaving the memory in `slice` unchanged.
+///
+/// # Precondition
+///
+/// If `slice.len() < Self::lanes()` or `&slice[0]` is not aligned
+/// to an `align_of::<Self>()` boundary, the behavior is undefined.
+pub unsafe fn read_aligned_unchecked(slice: &[element_type]) -> Self;
+
+/// Instantiates a new vector with the values of the `slice` without 
+/// moving them, leaving the memory in `slice` unchanged.
+///
+/// # Precondition
+///
+/// If `slice.len() < Self::lanes()` the behavior is undefined.
+pub unsafe fn read_unaligned_unchecked(slice: &[element_type]) -> Self;
+}
+```
+
+##### Discontinuous masked reads and writes (scatter and gather)
+
+Vector masks implement the following methods:
+
+```rust
+impl m{lane_width}x{number_of_lanes} {
+/// Instantiates a new vector with the values of the `slice` located at 
+/// the `offset`s without moving them for which the mask (`self`) is `true`
+/// and with the values of `default` otherwise. The memory of the `slice` at 
+/// the `offsets` for which the mask is `false` is not read.
+///
+/// # Precondition
+///
+/// If `slice.len() < offset.max_element()` the behavior is undefined.
+pub unsafe fn read_scattered_unchecked<T, O, D>(self, slice: &[T], offset: O, default: D) -> D
```

Comment:
This interface is not very useful, because it requires creating a safe slice that covers the entire region of memory one might want to read from. That means:

- you can't use one scatter to read from multiple distinct allocations at once
  - in particular, if you have a vector of *pointers* from some source and doesn't know anything about what they point at, you'd need a slice covering the entire address space which is obviously super UB
- even when working on one allocations, one can run into aliasing rules violations due to the slice covering elements you don't need access to (e.g., working on the first and third chunks of a bigger buffer, while another thread is working on the second chunk)
- you can't use it to read from partially-uninitialized or otherwise partially-invalid memory

Most of these issues are again due to taking a safe reference. The "one base pointer, vector of offsets" approach to gathers is salvagable but requires a lower level approach -- in rust terms, taking a raw pointer `base` and a vector of offsets, and then each lane reads `base.wrapping_offset(offsets.element(i))` or something like that.

Alternatively, I'd suggest the "vector of pointers" approach. It seems more intuitive and easier to get right. Do you know of a good reason not to?

Reply:
> This interface is not very useful,

Well, it covers a very common case of writing to multiple elements within an slice. It is not as general as supporting packed SIMD vectors of pointers. 

> Do you know of a good reason not to?

I tried a couple of times to implement vector of pointers but these are not currently easy to support due to many language limitations, IIRC: not being able to disjointly implement only one trait of a set of traits, e.g., a type implements either `Int|Uint|Float|Mask|Ptr|...` (cc @withoutboats), lack of trait `const fn` methods, ... 

The main issue is that vectors of pointers need to be generic types, because a vector of pointers to pointers to pointers to ... is also a very useful thing to have. This means that basically everything else needs to be generic to, because one needs to be able to read these vectors into another vector type, etc. 

I've just pushed a branch with the solution I got the furthest: https://github.com/gnzlbg/stdsimd/commit/24f6d34353f45afc37103e02f676327e30a6c672

Ping me on IRC if you want to discuss that more interactively, but basically I wanted these methods to cover the most common case of gather/scatters (which is a single indirection), without having to introduce vectors of pointers just yet. An alternative API that achieves the same thing but is more flexible would be to instead of taking a slice and a vector of offsets, to take an array of raw pointers instead and reinterpret that as a vector of pointers internally.